xla/service/scatter_determinism_expander.cc

```diff
@@ -0,0 +1,456 @@
+  auto* sorted_indices_following_part =
+      parent->AddInstruction(HloInstruction::CreateSlice(
+          ShapeUtil::MakeShape(scatter_indices->shape().element_type(),
+                               {indices_len - 1}),
+          sorted_scalar_indices, {1}, {indices_len}, {1}));
+  auto* indices_mask_without_padding =
+      parent->AddInstruction(HloInstruction::CreateCompare(
+          ShapeUtil::MakeShape(PRED, {indices_len - 1}),
+          sorted_indices_preceding_part, sorted_indices_following_part,
+          ComparisonDirection::kNe));
+  // Pad the comparison with a true value at the end
+  auto* true_constant = parent->AddInstruction(
+      HloInstruction::CreateConstant(LiteralUtil::CreateR0<bool>(true)));
+  auto* padding = parent->AddInstruction(HloInstruction::CreateBroadcast(
+      ShapeUtil::MakeShape(PRED, {1}), true_constant, {}));
+  std::vector<HloInstruction*> padding_operands = {indices_mask_without_padding,
+                                                   padding};
+  auto* indices_mask = parent->AddInstruction(HloInstruction::CreateConcatenate(
+      ShapeUtil::MakeShape(PRED, {indices_len}), padding_operands, 0));
+
+  // Mask the indices
+  indices_mask = parent->AddInstruction(HloInstruction::CreateBroadcast(
+      ShapeUtil::MakeShape(PRED, scatter_indices->shape().dimensions()),
+      indices_mask, {0}));
+
+  auto* out_of_bound_tensor =
+      CreateOutOfBoundTensor(parent, scatter_indices, scatter->shape());
+
+  auto* masked_indices = parent->AddInstruction(HloInstruction::CreateTernary(
+      sorted_indices->shape(), HloOpcode::kSelect, indices_mask, sorted_indices,
+      out_of_bound_tensor));
+  return masked_indices;
+}
+
+StatusOr<HloInstruction*> ScatterDeterminismExpander::ExpandInstruction(
+    HloInstruction* inst) {
+  auto* scatter = Cast<HloScatterInstruction>(inst);
+  auto scatter_operands = scatter->scatter_operands();
+  HloInstruction* scatter_indices = scatter->scatter_indices();
+  std::vector<HloInstruction*> scatter_updates(
+      scatter->scatter_updates().begin(), scatter->scatter_updates().end());
+  const ScatterDimensionNumbers& dim_numbers =
+      scatter->scatter_dimension_numbers();
+
+  // If the updates tensors are empty, there is no need to update the operands.
+  // The operands can be forwarded.
+  if (ShapeUtil::IsZeroElementArray(scatter_updates[0]->shape())) {
+    if (scatter_operands.size() == 1) {
+      return scatter_operands[0];
+    }
+    return scatter->parent()->AddInstruction(
+        HloInstruction::CreateTuple(scatter_operands));
+  }
+
+  // Compute the trip count for the while loop to be used for scatter. This
+  // should be the number of indices we should scatter into the operand.
+  int64_t scatter_indices_count = ScatterIndicesCount(scatter);
+  if (!IsInt32(scatter_indices_count)) {
+    // 2147483647 is the maximum value for a 32-bit signed integer (INT32_MAX).
+    return Unimplemented(
```

Comment:
I wonder whether it would better to just not match in this case (so moving the check to InstructionMatchesPattern), so that we can still support it (although very slow)?

Reply:
In the show version (`scatter_expander.cc`), they have exactly the same check and will abort so I followed that here. Even if we do not match this here, it will still abort in the `scatter_expander.cc`.

Reply:
Ah, that makes sense, thanks for the explanation.